src/content/1.7/modules/concepts/controllers/admin-controllers/override-decorate-controller.md

```diff
@@ -0,0 +1,153 @@
+
+So for example if you want that people browsing the Back Office on URL `/sell/orders/orders/` hit your own controller `MyModule\Controller\DemoController`, you can do this:
+```yaml
+# modules/your-module/config/routes.yml
+admin_orders_index:
+    path: /demo/orders
+    methods: [GET]
+    defaults:
+      _controller: 'MyModule\Controller\DemoController::demoAction'
+```
+
+{{% notice warning %}}
+Keep the items `_legacy_controller` and `_legacy_link` as they are necessary to enable matching with legacy URL `index.php?controller=AdminOrders` and modern URL `/sell/orders/orders/`
+{{% /notice %}}
+
+Thanks to this, whenever an HTTP request is matched to the route `admin_orders_index`, then your controller `demoAction()` will be executed.
+
+All module routes are prefixed with `modules`, which is why the URL of this remapped route is `modules/demo/orders`.
+
+{{% notice tip %}}
+Routing is computed and cached by Symfony. You will probably need to clear this cache for Symfony to acknowledge your updated routing.
+You can do it by using `php bin/console cache:clear`.
+{{% /notice %}}
+
+{{% notice tip %}}
+if you have trouble writing the right routing configuration for your controller, you can use Symfony debugger to dump the routes with `php bin/console debug:router`.
+{{% /notice %}}
+
+## Override the controller
+
+Since 1.7.7, PrestaShop Back Office controllers are registered as services, and services can be [overridden][override-services].
+
+For example, the controller responsible for the page "Improve > Design > CMS Pages" is registered with service ID `PrestaShopBundle\Controller\Admin\Improve\Design\CmsPageController`.
+
+With the following configuration item, we can override this configuration to make it target our custom module
+```yaml
+# modules/your-module/config/services.yml
+  'PrestaShopBundle\Controller\Admin\Improve\Design\CmsPageController':
+    class: MyModule\Controller\DemoController
+
+```
+
+Thanks to this, whenever Symfony forwards a request to the Core controller `PrestaShopBundle\Controller\Admin\Improve\Design\CmsPageController` it will be forwarded to `DemoController` instead.
+
+{{% notice tip %}}
+if you have trouble writing the right service configuration for your controller, you can use Symfony debugger to dump the routes with `php bin/console debug:container`. It can also be helpful to find the service ID of the controller.
+{{% /notice %}}
+
+## Decorate the controller
+
+However overriding a controller or a service is usually not recommanded, because this means you completely rewrite its behavior. If the implementation is updated in later versions of PrestaShop, your override will ignore these updates, which might create bugs.
+
+It is better to use [decoration](https://symfony.com/doc/3.4/service_container/service_decoration.html) which keeps the behavior of the decorated controller while allowing you to extend its behavior by modiying received inputs and outputs.
```

Comment:
```suggestion
Instead of replacing the whole controller, we recommend _extending_ its behavior using [service decoration](https://symfony.com/doc/3.4/service_container/service_decoration.html). By implementing the [decorator pattern](https://refactoring.guru/design-patterns/decorator), you can keep most or all of the original behavior of the decorated controller, and only customize the parts you want.

{{% notice note %}}
While you could achieve a similar end with an override (by making your controller extend the original one), the composition pattern provides a greater degree of freedom, while leaving all complexity of initialization and dependency management to the service container.
{{% /notice %}}
```